File: usecase-03/src/test/python/integration_test.py
import os
import sys
import unittest
import java
from PIL import Image, ImageChops
from simple_file_checksum import get_checksum

os.chdir("../../..")
testResourcesPath = 'src/test/resources/'
pyScriptsPath = 'src/main/python/'
sys.path.insert(0, os.path.abspath(pyScriptsPath))

import image_utils

class ImageUtilsTestCase(unittest.TestCase):
    def test_final_image_generation_is_successful(self):
        empty_img_path = image_utils.create_empty_image()
        qrcode_img_path = image_utils.create_qrcode_image("lorem ipsum dolor sit amet", "#A5FF00", "#5A5A5A")
        barcode_svg_img_path = image_utils.create_barcode_image("5909876183457")
        image_processor_class = java.type("com.yast.masterthesis.usecase03.ImageProcessor")
        barcode_png_img_path = image_processor_class.convertSvgToPng(os.path.abspath(barcode_svg_img_path))
        image_utils.rotate(barcode_png_img_path)
        out_img_path = image_utils.merge(empty_img_path, qrcode_img_path, 50, 50)
        out_img_path = image_utils.merge(out_img_path, barcode_png_img_path, 700, -60)
        image_utils.add_text(out_img_path, "Automatically generated!", 450, 230);
        image_utils.add_watermark(out_img_path, testResourcesPath + "stamp.png");
        test_img_path = testResourcesPath + "test_final_image.png"
        test_img = Image.open(test_img_path)
        out_img = Image.open(out_img_path)
        diff = ImageChops.difference(out_img, test_img)
        self.assertEqual(diff.getbbox(), None)
        checksum1 = get_checksum(out_img_path)
        checksum2 = get_checksum(test_img_path)
        self.assertEqual(checksum1, checksum2)

File: usecase-03/src/test/java/com/yast/masterthesis/usecase03/IntegrationTest.java
package com.yast.masterthesis.usecase03;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.security.DigestInputStream;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class IntegrationTest {
    @Test
    public void testFinalImageGenerationIsSuccessful() throws NoSuchAlgorithmException, IOException {
        String outImgPath = null;
        ImageProcessor imageProcessor = new ImageProcessor();
        String emptyImgPath = imageProcessor.createEmptyImage();
        String qrcodeImgPath = imageProcessor.createQrcodeImage("lorem ipsum dolor sit amet", "#A5FF00", "#5A5A5A");
        String barcodeSvgImgPath = imageProcessor.createBarcodeImage("5909876183457");
        String barcodePngImgPath = imageProcessor.convertSvgToPng(barcodeSvgImgPath);
        imageProcessor.rotateImage(barcodePngImgPath);
        outImgPath = imageProcessor.merge(emptyImgPath, qrcodeImgPath, 50, 50);
        outImgPath = imageProcessor.merge(outImgPath, barcodePngImgPath, 700, -60);
        imageProcessor.addText(outImgPath, "Automatically generated!", 450, 230);
        imageProcessor.addWatermark(outImgPath, "stamp.png");

        MessageDigest shaDigest = MessageDigest.getInstance("SHA-256");
        String checksum1 = checksum(outImgPath, shaDigest);
        String checksum2 = checksum("src/test/resources/test_final_image.png", shaDigest);
        assertEquals(checksum1, checksum2);
    }

    private static String checksum(String filepath, MessageDigest messageDigest) throws IOException {
        try (DigestInputStream dis = new DigestInputStream(new FileInputStream(filepath), messageDigest)) {
            while (dis.read() != -1) ;
            messageDigest = dis.getMessageDigest();
        }
        StringBuilder result = new StringBuilder();
        for (byte b : messageDigest.digest()) {
            result.append(String.format("%02x", b));
        }
        return result.toString();
    }
}
